src/sprightly.ts

```diff
@@ -1,60 +1,85 @@
-import { readFile } from "fs/promises"
+import { readFileSync } from "fs"
+import path from "path"
+import KeyNotFoundError from "./errors/KeyNotFoundError"
+import PathNotFoundError from "./errors/PathNotFoundError"
 
 interface Data {
   [key: string]: string | number | Array<string | number | Data> | Data
 }
 
-type Options = {
-  cache?: boolean
+interface Options {
   keyFallback?: string
   throwOnKeyNotfound?: boolean
+  cache?: boolean
 }
 
-const defaultOptions: Options = {
+const cache = new Map<string, string>()
+
+const defaultOptions = {
   keyFallback: "",
   throwOnKeyNotfound: false,
   cache: false,
 }
 
-export = async function sprightly(
+function parse(file: string, data: Data, options: Options) {
+  const parsedFile = file.replaceAll(
+    /\{\{(>?)(.*?)\}\}/g,
+    (_, isComponent: string, reference: string) => {
+      if (isComponent) {
+        try {
+          return sprightly(reference, data, options)
+        } catch (e) {
+          throw new PathNotFoundError(reference.trim())
+        }
+      }
+      return reference
+        .split(/((?:\w+))(?:(?:\[)(\d+)(?:\]))?(?:\.?)/gm)
+        .filter((v) => v && Boolean(v.trim()))
+        .reduce((obj, key) => {
+          if (
+            typeof obj === "object" &&
+            !(key in obj) &&
+            options.throwOnKeyNotfound
+          )
+            throw new KeyNotFoundError(key.trim())
+          return (obj[key] || options.keyFallback) as Data
+        }, data) as unknown as string
```

Comment:
This is actually not necessary. The final reduced file could still be an object.

Reply:
Maybe you need to handle this case where the "leaf" isn't a primitive value (string or number)?

Reply:
Got any ideas? or do you have anything specific in mind to print out if the reduced value is not a primitive?